src/ray/protobuf/instance_manager.proto

```diff
@@ -0,0 +1,98 @@
+syntax = "proto3";
+option cc_enable_arenas = true;
+
+package ray.rpc;
+
+// Represents the state of a launched instance.
+// An instance is considered launched as long as
+// it has a unique instance_id associated with it.
+//
+// Note a launched instance may be DEAD. In this case,
+// the state will be garbage collected after some timeout
+// period (by default 30 minutes).
+message LaunchedInstance {
+  enum LaunchedInstanceState {
+    // The instance is known to have finished booststrapping and
+    // should join the cluster soon/or have already
+    // joined the cluster.
+    ALIVE = 1;
+    // The instance is known to be dead. 
+    DEAD = 2;
+    // The node is going to be preempted by the node manager.
+    PREMEPTING = 3;
+    // The node is being terminated by the node provider.
+    TERMINATING = 4;
+    // The node has being allocated but envirnoment is not ready yet.
+    BOOSTRAPPING = 5;
+  }
+  bytes instance_id = 1;
+  string node_type = 2;
+  LaunchedInstanceState instance_state = 3; 
+}
+
+// Representing a pending request to launch an instance.
+//
+// Note a request could be failed due to various reasons.
+// The detailed reqson could be queried through description
+// field. A failed launching request will be GCed
+// after some timeout period (by default 30 minutes).
+message LauchingInstance {
+  enum LauchingRequestState {
+    // The request is queued and waiting to send.
+    QUEUED = 1;
+    // The request has been sent.
+    REQUESTED = 2;
+    // The request has been accepted, but no instance_id yet.
+    RESOURCES_ALLOCATED = 3;
+    // The request has failed.
+    FAILED = 4;
+  }
+  string node_type = 1;
+  string description = 2;
+  PendingRequestState pending_request_state = 3;
+}
+
+message GetInstanceManagerStateRequest {}
+
+message GetInstanceManagerStateReply {
+  // a monotonically increasing version number.
+  // the version number is incremented whenever
+  // the state is updated (either by successful adjusting request,
+  // or instance state change).
+  int64 version = 1;
+
+  // this number is increased whenever a successful
+  // adjusting request is received. Unlike the version,
+  // this doesn't change when instance state changes.
+  int64 num_successful_adjustments = 2;
```

Comment:
I don't really understand why this field is needed. What are the semantics if I send a duplicate request, a no-op request, or an invalid request?